File: ta4j-core/src/test/java/org/ta4j/core/rules/CrossedDownIndicatorRuleTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Before;
import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseBarSeriesBuilder;
import org.ta4j.core.indicators.helpers.FixedNumIndicator;

public class CrossedDownIndicatorRuleTest {

    private BarSeries series;

    @Before
    public void setUp() {
        series = new BaseBarSeriesBuilder().build();
    }

    @Test
    public void isSatisfied() {
        var evaluatedIndicator = new FixedNumIndicator(series, 12, 11, 10, 9, 11, 8, 7, 6);
        var rule = new CrossedDownIndicatorRule(evaluatedIndicator, 10);

        assertFalse(rule.isSatisfied(0));
        assertFalse(rule.isSatisfied(1));
        assertFalse(rule.isSatisfied(2));
        assertTrue(rule.isSatisfied(3));
        assertFalse(rule.isSatisfied(4));
        assertTrue(rule.isSatisfied(5));
        assertFalse(rule.isSatisfied(6));
        assertFalse(rule.isSatisfied(7));
    }

    @Test
    public void onlyThresholdBetweenFirstBarAndLastBar() {
        var evaluatedIndicator = new FixedNumIndicator(series, 11, 10, 10, 9);
        var rule = new CrossedDownIndicatorRule(evaluatedIndicator, 10);

        assertFalse(rule.isSatisfied(0));
        assertFalse(rule.isSatisfied(1));
        assertFalse(rule.isSatisfied(2));
        assertTrue(rule.isSatisfied(3));
    }

    @Test
    public void repeatedlyHittingThresholdAfterCrossDown() {
        var evaluatedIndicator = new FixedNumIndicator(series, 11, 10, 9, 10, 9, 10, 9);
        var rule = new CrossedDownIndicatorRule(evaluatedIndicator, 10);

        assertFalse(rule.isSatisfied(0));
        assertFalse(rule.isSatisfied(1));
        assertTrue("first cross down", rule.isSatisfied(2));
        assertFalse(rule.isSatisfied(3));
        assertFalse(rule.isSatisfied(4));
        assertFalse(rule.isSatisfied(5));
        assertFalse(rule.isSatisfied(6));
    }

    @Test
    public void serializeAndDeserialize() {
        var evaluatedIndicator = new FixedNumIndicator(series, 5, 6, 4, 3);
        var rule = new CrossedDownIndicatorRule(evaluatedIndicator, series.numFactory().numOf(5));
        RuleSerializationRoundTripTestSupport.assertRuleRoundTrips(series, rule);
        RuleSerializationRoundTripTestSupport.assertRuleJsonRoundTrips(series, rule);
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/rules/CrossedDownIndicatorRuleTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.ta4j.core.BarSeries
import org.ta4j.core.BaseBarSeries
import org.ta4j.core.Indicator
import org.ta4j.core.indicators.helpers.FixedDecimalIndicator
import org.ta4j.core.num.Num

class CrossedDownIndicatorRuleTest {
    private var rule: CrossedDownIndicatorRule? = null
    @Before
    fun setUp() {
        val series: BarSeries = BaseBarSeries()
        val evaluatedIndicator: Indicator<Num> =
            FixedDecimalIndicator(series, 12.0, 11.0, 10.0, 9.0, 11.0, 8.0, 7.0, 6.0)
        rule = CrossedDownIndicatorRule(evaluatedIndicator, 10)
    }

    @Test
    fun isSatisfied() {
        Assert.assertFalse(rule!!.isSatisfied(0))
        Assert.assertFalse(rule!!.isSatisfied(1))
        Assert.assertFalse(rule!!.isSatisfied(2))
        Assert.assertTrue(rule!!.isSatisfied(3))
        Assert.assertFalse(rule!!.isSatisfied(4))
        Assert.assertTrue(rule!!.isSatisfied(5))
        Assert.assertFalse(rule!!.isSatisfied(6))
        Assert.assertFalse(rule!!.isSatisfied(7))
    }
}